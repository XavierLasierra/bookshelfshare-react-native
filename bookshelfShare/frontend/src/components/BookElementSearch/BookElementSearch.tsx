import React from 'react';
import {
  View, Text, Image, TouchableOpacity
} from 'react-native';

import styles from './bookResults.styles';

export default function BookElementSearch({ bookData }: any) {
  return (
    <TouchableOpacity
      activeOpacity={0.9}
    >
      <Image
        source={{ uri: bookData?.images?.thumbnail || bookData?.images?.smallThumbnail }}
        style={styles.bookImage}
      />
      <View>
        <View>
          <View>
            <Text>{bookData?.title}</Text>
            <Text>{bookData?.language}</Text>
          </View>
          <Text>{bookData?.author}</Text>
          <Text>{bookData?.publisher}</Text>
        </View>
        <View>
          <Text>
            ISBN13 -
            {bookData?.isbn[0]?.identifier}
          </Text>
          <Text>
            ISBN10 -
            {bookData?.isbn[1]?.identifier}
          </Text>
        </View>
      </View>
    </TouchableOpacity>
  );
}
