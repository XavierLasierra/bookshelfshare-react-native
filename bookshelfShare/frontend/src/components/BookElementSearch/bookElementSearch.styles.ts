import {StyleSheet} from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  container: {
    flexDirection: 'row',
    marginVertical: 10,
    alignItems: 'center',
  },
  bookImage: {
    width: 125,
    height: 200,
    overflow: 'hidden',
    borderTopLeftRadius: 10,
    borderBottomLeftRadius: 10,
  },
  informationContainer: {
    backgroundColor: stylesConstants.colors.white,
    width: 210,
    height: 175,
    padding: stylesConstants.padding.small,
    borderTopRightRadius: 10,
    borderBottomRightRadius: 10,
    justifyContent: 'space-between',
  },
  titleLanguageContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    maxHeight: 75,
  },
  bookTitle: {
    width: 175,
    fontWeight: 'bold',
    fontSize: stylesConstants.fontSize.mid,
  },
  bookAuthor: {
    fontSize: stylesConstants.fontSize.small,
  },
  bookPublisher: {
    fontSize: stylesConstants.fontSize.small,
    color: stylesConstants.colors.mainText,
  },
  bookIsbn: {
    fontSize: stylesConstants.fontSize.small,
    color: stylesConstants.colors.mainText,
  },
});
