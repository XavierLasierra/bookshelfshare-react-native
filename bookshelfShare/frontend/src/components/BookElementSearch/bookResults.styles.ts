import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  bookImage: {
    width: 125,
    height: 200
  }
});
