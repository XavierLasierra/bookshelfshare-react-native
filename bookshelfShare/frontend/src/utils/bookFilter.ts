export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
