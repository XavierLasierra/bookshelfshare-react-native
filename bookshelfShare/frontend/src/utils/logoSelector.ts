import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
