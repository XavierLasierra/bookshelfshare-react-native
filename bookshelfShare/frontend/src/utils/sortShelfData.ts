export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
