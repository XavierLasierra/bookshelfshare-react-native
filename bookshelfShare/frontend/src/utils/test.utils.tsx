import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
