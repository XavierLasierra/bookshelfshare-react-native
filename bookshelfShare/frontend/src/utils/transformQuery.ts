interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
