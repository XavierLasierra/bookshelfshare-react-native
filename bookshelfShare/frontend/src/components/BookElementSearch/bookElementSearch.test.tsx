import React from 'react';
import BookElementSearch from './BookElementSearch';
import {render, fireEvent} from '../../utils/test.utils';
import bookDetailsMock from '../../mocks/bookDetails.mock';
import navigationMock from '../../mocks/navigation.mock';

describe('Given a BookElementSearch component', () => {
  describe('When it is rendered', () => {
    describe('And the book has a thumbnail and less than 3 authors', () => {
      let screen: any;
      let navigation: any;
      beforeEach(() => {
        navigation = {
          ...navigationMock,
          push: jest.fn(),
        };

        screen = render(
          <BookElementSearch
            bookData={bookDetailsMock[0]}
            navigation={navigation}
            logo="logo"
          />,
        );
      });

      test('Then should match the snapshot', () => {
        expect(screen).toMatchSnapshot();
      });

      describe('And bookDetail Button is clicked', () => {
        test('Then should call navigation.push with BookDetail and the book data', () => {
          const bookDetailButton = screen.getByTestId('bookDetailButton');
          fireEvent.press(bookDetailButton);
          expect(navigation.push).toHaveBeenCalledWith('BookDetail', {
            bookData: bookDetailsMock[0],
            logo: 'logo',
          });
        });
      });
    });
    describe('And the book has not a thumbnail and 3 authors or more', () => {
      test('Then should match the snapshot', () => {
        const screen = render(
          <BookElementSearch
            bookData={bookDetailsMock[1]}
            navigation={navigationMock}
            logo="logo"
          />,
        );
        expect(screen).toMatchSnapshot();
      });
    });
  });
});
