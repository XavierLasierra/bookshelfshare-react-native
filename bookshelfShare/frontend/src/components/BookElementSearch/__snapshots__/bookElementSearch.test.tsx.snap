// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a BookElementSearch component When it is rendered And the book has a thumbnail and less than 3 authors Then should match the snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "marginVertical": 10,
      "opacity": 1,
    }
  }
  testID="bookDetailButton"
>
  <Image
    source={
      Object {
        "uri": "thumbnail1",
      }
    }
    style={
      Object {
        "borderBottomLeftRadius": 10,
        "borderTopLeftRadius": 10,
        "height": 200,
        "overflow": "hidden",
        "width": 125,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomRightRadius": 10,
        "borderTopRightRadius": 10,
        "height": 175,
        "justifyContent": "space-between",
        "padding": 10,
        "width": 210,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "maxHeight": 75,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 16,
              "fontWeight": "bold",
              "width": 175,
            }
          }
        >
          1984
        </Text>
        <Text>
          EN
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        George Orwell
      </Text>
      <Text
        style={
          Object {
            "color": "#3E3C37",
            "fontSize": 12,
          }
        }
      >
        No publisher found
      </Text>
    </View>
    <View>
      <Text
        style={
          Object {
            "color": "#3E3C37",
            "fontSize": 12,
          }
        }
      >
        ISBN13 -
         
        9780451524935
      </Text>
      <Text
        style={
          Object {
            "color": "#3E3C37",
            "fontSize": 12,
          }
        }
      >
        ISBN10 -
         
        0451524934
      </Text>
    </View>
  </View>
</View>
`;

exports[`Given a BookElementSearch component When it is rendered And the book has not a thumbnail and 3 authors or more Then should match the snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "marginVertical": 10,
      "opacity": 1,
    }
  }
  testID="bookDetailButton"
>
  <Image
    source={
      Object {
        "uri": "smallThumbnail2",
      }
    }
    style={
      Object {
        "borderBottomLeftRadius": 10,
        "borderTopLeftRadius": 10,
        "height": 200,
        "overflow": "hidden",
        "width": 125,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomRightRadius": 10,
        "borderTopRightRadius": 10,
        "height": 175,
        "justifyContent": "space-between",
        "padding": 10,
        "width": 210,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "maxHeight": 75,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 16,
              "fontWeight": "bold",
              "width": 175,
            }
          }
        >
          Title
        </Text>
        <Text>
          EN
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Orwell
      </Text>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Another author
      </Text>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        ...
      </Text>
      <Text
        style={
          Object {
            "color": "#3E3C37",
            "fontSize": 12,
          }
        }
      >
        Debolsillo
      </Text>
    </View>
    <View>
      <Text
        style={
          Object {
            "color": "#3E3C37",
            "fontSize": 12,
          }
        }
      >
        ISBN13 -
         
        4567765432123
      </Text>
      <Text
        style={
          Object {
            "color": "#3E3C37",
            "fontSize": 12,
          }
        }
      >
        ISBN10 -
         
        0123456789
      </Text>
    </View>
  </View>
</View>
`;
